File: medium/String/find-all-anagrams-in-a-string/find_anagrams.py
from typing import List

def findAnagrams(s: str, p: str) -> List[int]:
    """
    Find all start indices of p's anagrams in s.
    Approach: Sliding Window + Frequency Count
    Time: O(n)
    Space: O(1)
    """
    len_s = len(s)
    len_p = len(p)
    res = []

    if len_p > len_s:
        return res

    count_s = [0] * 26
    count_p = [0] * 26

    # Initialize first window
    for i in range(len_p):
        count_p[ord(p[i]) - ord('a')] += 1
        count_s[ord(s[i]) - ord('a')] += 1

    if count_s == count_p:
        res.append(0)

    # Slide window
    for i in range(len_p, len_s):
        # Remove left
        count_s[ord(s[i - len_p]) - ord('a')] -= 1
        # Add right
        count_s[ord(s[i]) - ord('a')] += 1

        if count_s == count_p:
            res.append(i - len_p + 1)

    return res

if __name__ == "__main__":
    print(findAnagrams("cbaebabacd", "abc"))  # [0,6]

File: medium/String/find-all-anagrams-in-a-string/find_anagrams.cpp
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int len_s = s.size();
        int len_p = p.size();
        vector<int> res;

        if (len_p > len_s) return res;

        vector<int> count_s(26, 0);
        vector<int> count_p(26, 0);

        for (int i = 0; i < len_p; ++i) {
            count_p[p[i] - 'a']++;
            count_s[s[i] - 'a']++;
        }

        if (count_s == count_p) {
            res.push_back(0);
        }

        for (int i = len_p; i < len_s; ++i) {
            count_s[s[i - len_p] - 'a']--;
            count_s[s[i] - 'a']++;

            if (count_s == count_p) {
                res.push_back(i - len_p + 1);
            }
        }

        return res;
    }
};

// Test
#include <iostream>
int main() {
    Solution sol;
    vector<int> ans = sol.findAnagrams("cbaebabacd", "abc");
    for (int x : ans) cout << x << " ";
    return 0;
}
